specs_derive/src/lib.rs

```diff
@@ -0,0 +1,46 @@
+//! Custom `#[derive(Component)]` macro and `#[component(SomeStorageType)]`
+//! attribute for SPECS.
+
+extern crate proc_macro;
+#[macro_use]
+extern crate quote;
+extern crate syn;
+
+use proc_macro::TokenStream;
+use syn::{Ident, MacroInput, MetaItem, NestedMetaItem};
+use quote::Tokens;
+
+#[proc_macro_derive(Component, attributes(component))]
+pub fn component(input: TokenStream) -> TokenStream {
+    let s = input.to_string();
+    let ast = syn::parse_derive_input(&s).unwrap();
+    let gen = impl_component(&ast);
+    gen.parse().unwrap()
+}
+
+fn impl_component(ast: &MacroInput) -> Tokens {
+    let name = &ast.ident;
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
```

Comment:
It's very nice! I learned about it from the [Procedural Macros chapter](https://doc.rust-lang.org/beta/book/first-edition/procedural-macros.html) of the Rust book.